File: agenkit-scala/src/main/scala/io/agenkit/patterns/TaskAgent.scala
package io.agenkit.patterns

import io.agenkit.adapters.LlmClient
import io.agenkit.core.*
import scala.concurrent.{ExecutionContext, Future}
import java.util.concurrent.atomic.{AtomicLong, AtomicReference}

enum TaskStatus:
  case Pending, Running, Complete, Failed

/** Single-task agent that tracks its own lifecycle status. */
class TaskAgent(
  val name: String,
  llm: LlmClient
) extends Agent:
  private val _status          = new AtomicReference[TaskStatus](TaskStatus.Pending)
  private val _processedCount  = new AtomicLong(0)

  def status: TaskStatus = _status.get()

  def capabilities: List[String] = List("task", "lifecycle")

  def process(message: Message)(using ExecutionContext): Future[Message] =
    _status.set(TaskStatus.Running)
    _processedCount.incrementAndGet()
    llm.complete(List(message))
      .map { response =>
        _status.set(TaskStatus.Complete)
        response
      }
      .recoverWith { case ex =>
        _status.set(TaskStatus.Failed)
        Future.failed(ex)
      }

  def introspect(): IntrospectionResult =
    IntrospectionResult(
      name = name,
      metadata = Map("status" -> _status.get().toString),
      capabilities = capabilities,
      processedMessages = _processedCount.get()
    )

File: agenkit-java/src/main/java/io/agenkit/patterns/TaskAgent.java
package io.agenkit.patterns;

import io.agenkit.adapters.LlmClient;
import io.agenkit.core.Agent;
import io.agenkit.core.IntrospectionResult;
import io.agenkit.core.Message;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicReference;

/**
 * A one-shot task agent with explicit lifecycle management.
 */
public final class TaskAgent implements Agent {

    public enum Status { IDLE, RUNNING, COMPLETE, FAILED }

    private final String name;
    private final LlmClient llmClient;
    private final String taskDescription;
    private final AtomicReference<Status> status = new AtomicReference<>(Status.IDLE);
    private volatile Message lastResult;

    public TaskAgent(String name, LlmClient llmClient, String taskDescription) {
        this.name = name;
        this.llmClient = llmClient;
        this.taskDescription = taskDescription;
    }

    @Override
    public String getName() { return name; }

    @Override
    public List<String> getCapabilities() {
        return List.of("task_execution", "lifecycle_management");
    }

    @Override
    public CompletableFuture<Message> process(Message message) {
        if (!status.compareAndSet(Status.IDLE, Status.RUNNING)) {
            return CompletableFuture.completedFuture(
                    Message.of("assistant", "Task agent is not idle. Status: " + status.get()));
        }

        return llmClient.complete(List.of(
                Message.of("system", "Task: " + taskDescription + "\nComplete this specific task."),
                message))
                .thenApply(response -> {
                    status.set(Status.COMPLETE);
                    lastResult = response;
                    return response.withMetadata("task_status", "complete")
                            .withMetadata("task", taskDescription);
                })
                .exceptionally(ex -> {
                    status.set(Status.FAILED);
                    Message err = Message.of("assistant", "Task failed: " + ex.getMessage())
                            .withMetadata("task_status", "failed");
                    lastResult = err;
                    return err;
                });
    }

    public void reset() { status.set(Status.IDLE); }

    public Status getStatus() { return status.get(); }

    public Message getLastResult() { return lastResult; }

    @Override
    public IntrospectionResult introspect() {
        Map<String, Object> state = new HashMap<>();
        state.put("status", status.get().name());
        state.put("task", taskDescription);
        return new IntrospectionResult(name, getCapabilities(), null, state, null);
    }
}
